### PR TITLE
Fix #271505: Allow user to set mode along with key signature

### PR DESCRIPTION
**Pull request by Jojo-Schmitz**
MusicXML im- and export of all possible key signatur modes.
Make these availble to the UI, via the Key Signature Inspector. (WIP, it doesn't quite work yet, any idea why?)

### CODE REVIEW

**Comment by marnen** [comment]
Oh awesome. Thanks!

File: libmscore/keysig.cpp [hunk 1]
```diff
@@ -510,14 +531,24 @@ void KeySig::undoSetShowCourtesy(bool v)
       undoChangeProperty(Pid::SHOW_COURTESY, v);
       }
 
+//---------------------------------------------------------
+//   undoSetModeType
+//---------------------------------------------------------
+
+void KeySig::undoSetModeType(KeyMode v)
```
Comment (lasconic) [inline libmscore/keysig.cpp]: Why modetype? and not just mode ?

**Comment by lasconic** [comment]
What part is no working specifically? How do you plan to change the mode of C major ?

File: libmscore/keysig.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [reply libmscore/keysig.cpp]: I can change that, no problem

**Comment by Jojo-Schmitz** [comment]
The dialog in Inspector doesn't work, it doesn't show the mode and doesn't allow to change it

File: libmscore/keysig.cpp [hunk 2]
```diff
@@ -535,6 +566,11 @@ bool KeySig::setProperty(Pid propertyId, const QVariant& v)
                         return false;
                   setShowCourtesy(v.toBool());
                   break;
+            case Pid::KEYSIG_MODETYPE:
+                  if (generated())
+                        return false;
+                  //setModeType(v);
```
Comment (lasconic) [inline libmscore/keysig.cpp]: This could be the reason why it's not set by the inspector maybe ?

File: mscore/inspector/inspector.cpp [hunk 3]
```diff
@@ -678,7 +678,8 @@ InspectorKeySig::InspectorKeySig(QWidget* parent)
       const std::vector<InspectorItem> iiList = {
             { Pid::LEADING_SPACE,  1, s.leadingSpace,  s.resetLeadingSpace  },
             { Pid::SHOW_COURTESY,  0, k.showCourtesy,  k.resetShowCourtesy  },
-//          { Pid::SHOW_NATURALS,  0, k.showNaturals,  k.resetShowNaturals  }
+          //{ Pid::SHOW_NATURALS,  0, k.showNaturals,  k.resetShowNaturals  },
+            { Pid::KEYSIG_MODETYPE,0, k.modeType,      k.resetModeType      }
```
Comment (lasconic) [inline mscore/inspector/inspector.cpp]: This is not enough to make the combo works. You need to define user data in the combo to match the value of the Keymode enum. If not, MuseScore uses enum value as index `void InspectorBase::setValue(const InspectorItem& ii, QVariant val)`

File: libmscore/keysig.cpp (hunk 2 above)
Comment (Jojo-Schmitz) [reply libmscore/keysig.cpp]: It is commented out because it doesn't compile ;-)

File: mscore/inspector/inspector.cpp (hunk 3 above)
Comment (Jojo-Schmitz) [reply mscore/inspector/inspector.cpp]: OK, this might be it then.

File: libmscore/keysig.cpp (hunk 2 above)
Comment (Jojo-Schmitz) [reply libmscore/keysig.cpp]: no known conversion from 'const QVariant' to 'Ms::KeyMode'

File: libmscore/keysig.cpp [hunk 4]
```diff
@@ -535,6 +566,11 @@ bool KeySig::setProperty(Pid propertyId, const QVariant& v)
                         return false;
                   setShowCourtesy(v.toBool());
                   break;
+            case Pid::KEYSIG_MODE:
+                  if (generated())
+                        return false;
+                  //setMode(v);  // no known conversion from 'const QVariant' to 'Ms::KeyMode'
```
Comment (lasconic) [inline libmscore/keysig.cpp]: KeyMode(v.toInt()) ?

Comment (Jojo-Schmitz) [reply libmscore/keysig.cpp]: error: no matching function for call to 'Ms::KeySig::setMode(int)'

Comment (Jojo-Schmitz) [reply libmscore/keysig.cpp]: oops, indeed

File: mscore/inspector/inspector.cpp (hunk 3 above)
Comment (Jojo-Schmitz) [reply mscore/inspector/inspector.cpp]: Now at least that menu forsn't start empty, but with "Unknown", still the setting doesn't stick though

**Comment by Jojo-Schmitz** [comment]
note to myself: rebase needed

File: libmscore/property.cpp [hunk 5]
```diff
@@ -660,6 +661,18 @@ QString propertyToString(Pid id, QVariant value, bool mscx)
...
                   return value.toString();
+            case P_TYPE::KEYMODE:
+                  switch (KeyMode(value.toInt())) {
+                        case KeyMode::NONE:
+                              return "none";
+                        case KeyMode::MAJOR:
+                              return "major";
+                        case KeyMode::MINOR:
+                              return "minor";
+                        default:
```
Comment (dmitrio95) [inline libmscore/property.cpp]: This `switch` should probably include all the new `KeyMode` entries too (if this PR is still actual).

Comment (Jojo-Schmitz) [reply libmscore/property.cpp]: oops

**Comment by Jojo-Schmitz** [comment]
Strange, why do I get merge conflicts on rebase, but none are reported here?

File: libmscore/property.cpp (hunk 5 above)
Comment (Jojo-Schmitz) [reply libmscore/property.cpp]: fixed

**Comment by Jojo-Schmitz** [comment]
Ooops, something went wrong tit inspector_keysig.ui...

File: mscore/inspector/inspector.cpp [hunk 6]
```diff
@@ -797,20 +797,35 @@ InspectorKeySig::InspectorKeySig(QWidget* parent)
...
 
 void InspectorKeySig::setElement()
       {
       InspectorElementBase::setElement();
       KeySig* ks = toKeySig(inspector->element());
-      if (ks->generated())
+      if (ks->generated()) {
             k.showCourtesy->setEnabled(false);
+            k.keysigModeLabel->setEnabled(false);
+            k.keysigMode->setEnabled(false);
```
Comment (Jojo-Schmitz) [inline mscore/inspector/inspector.cpp]: Not sure about those, shouldn't the initial (and as such 'generated') timesig allow that setting too?

Comment (Jojo-Schmitz) [reply mscore/inspector/inspector.cpp]: Forget it, it is already...

**Comment by Jojo-Schmitz** [comment]
@anatoly-os I beliebe the "work in progress" label can get removed

**Pull request merged**